SshAgentLib/UnixSocket.cs

```diff
@@ -0,0 +1,221 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+using Microsoft.Win32.SafeHandles;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Security.AccessControl;
+using System.Security.Principal;
+using System.Text;
+using System.Threading;
+
+namespace dlech.SshAgentLib
+{
+  public class UnixSocket : IDisposable
+  {
+    const string waitHandleNamePrefix = "unix.local_socket.secret";
+
+    static int clientCount = 0;
+
+    string path;
+    Socket socket;
+    SocketAddress sockaddr;
+    Guid guid;
+    bool disposed;
+    List<Socket> clientSockets = new List<Socket>();
+    object clientSocketsLock = new object();
+
+    public delegate void ConnectionHandlerFunc(Stream stream, Process process);
+    public ConnectionHandlerFunc ConnectionHandler { get; set; }
+
+    /// <summary>
+    /// Create new "unix domain" socket for use with Linux
+    /// </summary>
+    /// <param name="path">The name of the file to use for the socket</param>
+    public UnixSocket(string path)
+    {
+      this.path = path;
+      guid = Guid.NewGuid();
+      {
+        try {
+          socket = new Socket(AddressFamily.Unix, SocketType.Stream,
+            ProtocolType.Unspecified);
+          var endpoint = new UnixEndPoint(path);
+          sockaddr = endpoint.Serialize();
+          socket.Bind(endpoint);
+          var fileSecurity = File.GetAccessControl(path);
+          // This turns off ACL inheritance and removes all inherited rules
+          fileSecurity.SetAccessRuleProtection(true, false);
+          // We are left with no permissions at all, so we have to add them
+          // back for the current user
+          var userOnlyRule = new FileSystemAccessRule(
+            WindowsIdentity.GetCurrent().User,
+            FileSystemRights.FullControl,
+            AccessControlType.Allow);
+          fileSecurity.SetAccessRule(userOnlyRule);
+          File.SetAccessControl(path, fileSecurity);
+          socket.Listen(5);
+          var socketThread = new Thread(AcceptConnections);
+          socketThread.Name = "UnixSocket";
+          socketThread.Start();
+        } catch (Exception) {
+          if (socket != null)
+            socket.Close();
+          File.Delete(path);
+          throw;
+        }
+      }
+    }
+
+    /// <summary>
+    /// Tests a file to see if it looks like a Unix socket file
+    /// </summary>
+    /// <param name="path">The path to the file.</param>
+    /// <returns><c>true</c> if the file contents look correct</returns>
+    public static bool TestFile(string path)
```

Comment:
I think the name is not very clear. Maybe this instead?

```suggestion
    public static bool IsUnixSocket(string path)
```

Reply:
I copied the name over from `MsysSocket.cs` and `CygwinSocket.cs`. Might be worth changing if you also change those at some point?